### PR TITLE
Direction dimension discretization

### PR DESCRIPTION
**Pull request by psbritt**
Minus PyFrensie interface and archiving (continue run) capability for this feature addition, direction discretization for observers (estimators, etc.) have been implemented and tested.

### CODE REVIEW

File: packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  void calculateBinIndicesOfValue(  const boost::any& any_value,
+                                            BinIndexArray& bin_indices ) final override;
+
+    //! Calculate the index of bins that the value range falls in
+  void calculateBinIndicesOfRange( const ObserverParticleStateWrapper& particle_state_wrapper,
+                                  BinIndexWeightPairArray& bin_indices_and_weights )final override;
+
+  //! Print the boundaries of a bin
+  void printBoundariesOfBin( std::ostream& os,
+				                             const size_t bin_index )final override;
```
Comment (bam241) [inline packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp:54]: it seems to you are mixing tabs and spaces for indentation.

I would strongly suggest to give up on tabs :)

File: packages/monte_carlo/event/core/src/MonteCarlo_ObserverDirectionDimensionDiscretization.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  //! calculate bin index for direction
+  virtual void calculateBinIndicesOfValue( const ObserverParticleStateWrapper& particle_state_wrapper,
+                                          BinIndexArray& bin_indices) const = 0;
+                                      
+  //! Calculate the index of bins that the value falls in
+  virtual void calculateBinIndicesOfValue( const ObserverParticleStateWrapper& particle_state_wrapper,
+                                          BinIndexWeightPairArray& bin_indices_and_weights ) const = 0;
+
+    //! Calculate the index of bins that the value falls in
+  virtual void calculateBinIndicesOfValue(  const boost::any& any_value,
```
Comment (bam241) [inline packages/monte_carlo/event/core/src/MonteCarlo_ObserverDirectionDimensionDiscretization.hpp:61]: 2 space after `(` convention seems to be only 1

File: packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  if( d_forward_bin)
+  {  bin_indices[0] = d_pqla_quadrature_handler.findTriangleBin( particle_state_wrapper.getParticleState().getXDirection(),
+                                                                 particle_state_wrapper.getParticleState().getYDirection(),
+                                                                 particle_state_wrapper.getParticleState().getZDirection() );
+  }else if(!d_forward_bin)
+  {
+    bin_indices[0] = d_pqla_quadrature_handler.findTriangleBin( -particle_state_wrapper.getParticleState().getXDirection(),
+                                                                -particle_state_wrapper.getParticleState().getYDirection(),
+                                                                -particle_state_wrapper.getParticleState().getZDirection() );
+  }else
```
Comment (bam241) [inline packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.cpp:41]: if `d_foward_bin` is a boolean, I doubt It is possible to reach the exception

File: packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  }else if( !d_forward_bin )
+  {
+    bin_indices_and_weights[0].first = d_pqla_quadrature_handler.findTriangleBin( -particle_state_wrapper.getParticleState().getXDirection(),
+                                                                                  -particle_state_wrapper.getParticleState().getYDirection(),
+                                                                                  -particle_state_wrapper.getParticleState().getZDirection() );
+  }else
+  {
+    THROW_EXCEPTION( std::logic_error,
+                     "Standard/Reverse direction binning not defined" );
+  }
```
Comment (bam241) [inline packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.cpp:71]: I would try to factories this piece of code with the one above (l-23 to l45 and l57 to l71)
with something like get `quadrature_handler` returning the handler and dealing with the exception if need be...

File: packages/utility/direction_discretization/src/Utility_PQLAQuadrature.cpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+  }else if( !x_sign && y_sign && z_sign)
+  {
+
+    return 6;
+
+  }else if( x_sign && y_sign && z_sign)
+  {
+
+    return 7;
```
Comment (bam241) [inline packages/utility/direction_discretization/src/Utility_PQLAQuadrature.cpp:147]: if the secondary index value order is just a convention
if might be more elegant to do something like:
```
int idx = 0
if (x_sign) idx +=1
if (y_sign) idx +=2
if (x_sign) idx +=4

return idx
```
this should give you unique values ranging between 0 and 7  based on x_sign/y_sign/x_sign true/false values


I am still not understand what bring the exception there.

File: packages/utility/direction_discretization/test/tstPQLAQuadrature.cpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+  direction[0] = 1;
+  direction[1] = 2;
+  direction[2] = 0;
+
+  Utility::normalizeVector(direction[0],
+                           direction[1],
+                           direction[2]); 
+
+  FRENSIE_CHECK_EQUAL( PQLAQuadrature->findTriangleBin(direction), 5);
```
Comment (bam241) [inline packages/utility/direction_discretization/test/tstPQLAQuadrature.cpp:122]: because of all the blank lines it is hard to catch the meaning-full code block.
I would suggest to only have empty blank line between the different tests

File: packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
+//---------------------------------------------------------------------------//
+//!
+//! \file   MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.hpp
+//! \author Philip Britt
+//! \brief  PQLA Direction Discretization declaration
+//!
+//---------------------------------------------------------------------------//
+
+#ifndef MONTE_CARLO_PQLA_TYPE_OBSERVER_DIRECTION_DIEMNSION_DISCRETIZATION
+#define MONTE_CARLO_PQLA_TYPE_OBSERVER_DIRECTION_DIEMNSION_DISCRETIZATION
```
Comment (bam241) [inline packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp:10]: there is a typo in the variable

Comment (bam241) [inline packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp:10]: Those import guard are used 2 times. (identical guards are used in 1MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.hpp `)

I am doubting this is actually compiled.

File: packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
+//---------------------------------------------------------------------------//
```
Comment (bam241) [inline packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp:1]: this file as no corresponding cpp

I don't fully understand the use/need of this emptyType.
But I am pretty sure it is not used/compiled right now  (import guard conflict and no associated cpp file)


this would probably not work well either as some method are not implemented.

File: packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp (hunk 7 above)
Comment (psbritt) [reply packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp:10]: The empty type observer direction discretization class files were supposed to be deleted. I'm not sure how it made it into the PR?

File: packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.cpp (hunk 3 above)
Comment (psbritt) [reply packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.cpp:41]: I agree with you, but these exceptions are just to give me peace of mind in case there's something I didn't know/didn't fully understand about C++ so if an error like that is hit in these methods, I don't end up spending a lot of time trying to figure it out and know exactly where the error is.

File: packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.cpp (hunk 4 above)
Comment (psbritt) [reply packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.cpp:71]: I don't quite understand.

File: packages/utility/direction_discretization/test/tstPQLAQuadrature.cpp (hunk 6 above)
Comment (psbritt) [reply packages/utility/direction_discretization/test/tstPQLAQuadrature.cpp:122]: These blank lines are generally a formatting style that's used throughout FRENSIE, I believe. I don't think I should remove them and make my tests different style-wise from the others

File: packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.cpp (hunk 4 above)
Comment (bam241) [reply packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.cpp:71]: basically implement a something like:
```
get_handler(particle_state_wrapper){

if( d_forward_bin )
  {  
    return  d_pqla_quadrature_handler.findTriangleBin( particle_state_wrapper.getParticleState().getXDirection(),
                                                                                   particle_state_wrapper.getParticleState().getYDirection(),
                                                                                   particle_state_wrapper.getParticleState().getZDirection() );
  }else
  {
    return d_pqla_quadrature_handler.findTriangleBin( -particle_state_wrapper.getParticleState().getXDirection(),
                                                                                  -particle_state_wrapper.getParticleState().getYDirection(),
                                                                                  -particle_state_wrapper.getParticleState().getZDirection() );
  }
}
```
and then inside both methods call:

`bin_indices_and_weights[0].first = get_handler()`
or
`bin_indices[0] = get_handler()`

File: packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp (hunk 7 above)
Comment (psbritt) [reply packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp:10]: They were deleted, this file is just an incorrectly named one that got included by accident. It should be removed soon.

File: packages/utility/direction_discretization/src/Utility_PQLAQuadrature.cpp [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+  // Add the basic equation for calculating the triangle index
+  sum = sum + d_quadrature_order + i_z - i_x - i_y - 1;
+
+  // handle edge cases, default to lower plane index on edge cases (i_z is added, i_y is subtracted, making the below logic correct)
+  if( i_x + i_y + i_z == d_quadrature_order)
+  {
+    if( i_z > 0 )
+    {
+      sum = sum - 1;
+    }else if ( i_x == d_quadrature_order)
```
Comment (gonuke) [inline packages/utility/direction_discretization/src/Utility_PQLAQuadrature.cpp:95]: Why not change this condition and only keep the case where there is action?

File: packages/utility/direction_discretization/test/tstPQLAQuadrature.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+  std::shared_ptr<Utility::PQLAQuadrature> PQLAQuadrature(
+              new Utility::PQLAQuadrature( 3 ) );
+
+  std::array<double, 3> direction {1, 1, 2};
+
+  Utility::normalizeVector(direction[0],
+                  direction[1],
+                  direction[2]);
+  FRENSIE_CHECK_EQUAL( PQLAQuadrature->findTriangleBin(direction),  3);
+
```
Comment (gonuke) [inline packages/utility/direction_discretization/test/tstPQLAQuadrature.cpp:56]: It might be helpful for future authors/developers to note which of these are edge cases.  Or possibly put edge cases in their own tests.

File: packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+//!
+//! \file   MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.hpp
```
[the remark this answers is not in the document]
Comment (psbritt) [reply packages/monte_carlo/event/core/src/MonteCarlo_MonteCarlo_EmptyTypeObserverDirectionDimensionDiscretization.hpp]: This file was named incorrectly which caused me to miss it when deleting files. The class was no longer needed so it was deleted.

File: packages/monte_carlo/event/core/src/MonteCarlo_ObserverDirectionDimensionDiscretization.hpp [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+                                            BinIndexArray& bin_indices ) const = 0;
+
+    //! Calculate the index of bins that the value range falls in
+  virtual void calculateBinIndicesOfRange( const ObserverParticleStateWrapper& particle_state_wrapper,
+                                  BinIndexWeightPairArray& bin_indices_and_weights ) const = 0;
+
+  //! Print the boundaries of a bin
+  virtual void printBoundariesOfBin( std::ostream& os,
+				                             const size_t bin_index ) const = 0;
```
[the remark this answers is not in the document]
Comment (psbritt) [reply packages/monte_carlo/event/core/src/MonteCarlo_ObserverDirectionDimensionDiscretization.hpp:70]: I don't understand what's going on here. I'm not seeing this white space on my machine, and it doesn't look like the last commit changed this.

[the remark this answers is not in the document]
Comment (bam241) [reply packages/monte_carlo/event/core/src/MonteCarlo_ObserverDirectionDimensionDiscretization.hpp:70]: it is because you are using tabs....

they render well on you editor because the editor represents as a certain number of spaces. Github is probably representing space as a different number of spaces...
better to use x number spaces over tabs (a common value we used on other project are 2 spaces for 1 tab... -- for C++ file, 4 spaces for a tab in python...)

File: packages/monte_carlo/event/core/src/MonteCarlo_ObserverPhaseSpaceDimension.cpp [hunk 13]
```diff
@@ -36,6 +36,8 @@ std::string ToStringTraits<MonteCarlo::ObserverPhaseSpaceDimension>::toString(
       return MonteCarlo::ObserverPhaseSpaceDimensionTraits<MonteCarlo::OBSERVER_COLLISION_NUMBER_DIMENSION>::name();
     case MonteCarlo::OBSERVER_SOURCE_ID_DIMENSION:
       return MonteCarlo::ObserverPhaseSpaceDimensionTraits<MonteCarlo::OBSERVER_SOURCE_ID_DIMENSION>::name();
+    case MonteCarlo::OBSERVER_DIRECTION_DIMENSION:
+      return "Direction";
```
[the remark this answers is not in the document]
Comment (psbritt) [reply packages/monte_carlo/event/core/src/MonteCarlo_ObserverPhaseSpaceDimension.cpp:40]: There is one, but wouldn't I have to instantiate the object in order to use it? It's not instantiated by the member data (d_impl), so wouldn't that be unnecessary memory overhead to instantiate it here for a single method?

**Comment by psbritt** [comment]
I think I've made all of the changes based on comments received so far.

File: packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.hpp [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+//! \brief  PQLA Direction Discretization declaration
+//!
+//---------------------------------------------------------------------------//
+
+#ifndef MONTE_CARLO_PQLA_TYPE_OBSERVER_DIRECTION_DIMENSION_DISCRETIZATION
+#define MONTE_CARLO_PQLA_TYPE_OBSERVER_DIRECTION_DIMENSION_DISCRETIZATION
+
+// FRENSIE includes
+#include "MonteCarlo_ObserverDirectionDimensionDiscretization.hpp"
+#include "Utility_PQLAQuadrature.hpp"
```
Comment (gonuke) [inline packages/monte_carlo/event/core/src/MonteCarlo_PQLATypeObserverDirectionDimensionDiscretization.hpp:14]: Does this need to be included in the header, or can you just forward declare the class and include in the implementation?

File: packages/CMakeLists.txt [hunk 15]
```diff
@@ -17,7 +17,7 @@ SET(GLOBAL_ROOT_TEST_DATA_BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/test_files/root
 # Set up the packages directory hierarchy
 ADD_SUBDIRECTORY(utility)
 INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR}/utility/core/src
-utility/core/src utility/archive/src utility/archive/test utility/system/src utility/mpi/src utility/prng/src utility/interpolation/src utility/distribution/src utility/stats/src utility/integrator/src utility/grid/src utility/mesh/src)
+utility/core/src utility/archive/src utility/archive/test utility/system/src utility/mpi/src utility/prng/src utility/interpolation/src utility/direction_discretization/src utility/distribution/src utility/stats/src utility/integrator/src utility/grid/src utility/mesh/src)
```
[the remark this answers is not in the document]
Comment (aprobinson) [reply packages/CMakeLists.txt:20]: Definitely my fault there. @psbritt CMake doesn't care about newlines. We could put one directory per line instead of only using a single line to make it easier to read

File: packages/monte_carlo/event/estimator/test/tstHexMeshTrackLengthFluxEstimator.cpp [hunk 16]
```diff
@@ -118,6 +118,15 @@ FRENSIE_UNIT_TEST_TEMPLATE( HexMeshTrackLengthFluxEstimator,
   FRENSIE_CHECK_EQUAL( estimator->getNumberOfBins( MonteCarlo::OBSERVER_COSINE_DIMENSION ),
                        1 );
```
[the remark this answers is not in the document]
Comment (aprobinson) [reply packages/monte_carlo/event/estimator/test/tstHexMeshTrackLengthFluxEstimator.cpp:119]: If it was a surface estimator, yes. Cosine binning isn't allowed for the cell/mesh based estimators so they are simply ignored (with a warning message). This check verifies that they have indeed been ignored.

File: packages/utility/direction_discretization/src/Utility_PQLAQuadrature.hpp [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+  //! Find which triangle bin a direction vector is in
+  unsigned findTriangleBin( const double x_direction, const double y_direction, const double z_direction) const;
+
+  //! Return the order of the quadrature
+  unsigned getQuadratureOrder() const;
+
+  private:
+
+  //! Converts direction vector to 1-norm normalized vector
+  std::array<double, 3> normalizeVectorToOneNorm( std::array<double, 3> direction_2_norm) const;
```
Comment (aprobinson) [inline packages/utility/direction_discretization/src/Utility_PQLAQuadrature.hpp:43]: There is potentially some unnecessary copy overhead with this return value since it's an array of three doubles. This isn't a huge deal as the array is statically sized and small but it would be more efficient to pass it in by reference. The same goes for your input parameter (it should be "const std::array<double, 3>& direction_2_norm" to avoid copy overhead (does std::array do deep copies or shallow copies?).

File: packages/utility/direction_discretization/src/Utility_PQLAQuadrature.hpp [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+  
+  //! Converts direction vector to 1-norm normalized vector
+  std::array<double, 3> normalizeVectorToOneNorm(double x_direction, double y_direction, double z_direction) const;
```
Comment (aprobinson) [inline packages/utility/direction_discretization/src/Utility_PQLAQuadrature.hpp:46]: see my previous comment about return value copy overhead

**Comment by psbritt** [comment]
I think I've made all changes. Can this be merged?

**Comment by gonuke** [comment]
I'd like to let @aprobinson have final say on this, until we are all a little more familiar....

File: packages/utility/direction_discretization/src/Utility_PQLAQuadrature.hpp [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+  //! Find which triangle bin a direction vector is in
+  unsigned findTriangleBin( const double x_direction, const double y_direction, const double z_direction) const;
+
+  //! Return the order of the quadrature
+  unsigned getQuadratureOrder() const;
+
+  private:
+
+  //! Converts direction vector to 1-norm normalized vector
+  void normalizeVectorToOneNorm( const std::array<double, 3> direction_2_norm,
```
Comment (aprobinson) [inline packages/utility/direction_discretization/src/Utility_PQLAQuadrature.hpp:43]: I think you forgot the reference after 'const std::array<double, 3>`

File: packages/utility/direction_discretization/src/Utility_PQLAQuadrature.cpp [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+  return this->findTriangleBin(direction_array);
+}
+
+unsigned PQLAQuadrature::getQuadratureOrder() const
+{
+  return d_quadrature_order;
+}
+
+// Converts direction vector to 1-norm normalized vector
+void PQLAQuadrature::normalizeVectorToOneNorm( const std::array<double, 3> direction_normalized_2_norm,
```
Comment (aprobinson) [inline packages/utility/direction_discretization/src/Utility_PQLAQuadrature.cpp:60]: I think you forgot the reference after 'const std::array<double, 3>'

File: packages/utility/direction_discretization/src/Utility_PQLAQuadrature.cpp [hunk 21]
```diff
@@ -44,7 +44,7 @@ unsigned PQLAQuadrature::findTriangleBin(const std::array<double, 3>& direction)
 }  
 
 // Find which triangle bin a direction vector is in (takes 2-norm vector)
-unsigned PQLAQuadrature::findTriangleBin( const double x_direction, const double y_direction, const double z_direction) const
+unsigned PQLAQuadrature::findTriangleBin( const double& x_direction, const double& y_direction, const double& z_direction) const
```
Comment (gonuke) [inline packages/utility/direction_discretization/src/Utility_PQLAQuadrature.cpp:47]: These (and some of the others) are probably less important... this doesn't change efficiency at all - instead of passing a single 8 byte double it passes a single 8 byte pointer....

**Comment by psbritt** [comment]
Is there no efficiency change due to the difference in copying the variable?

File: packages/utility/direction_discretization/src/Utility_PQLAQuadrature.cpp (hunk 21 above)
Comment (gonuke) [reply packages/utility/direction_discretization/src/Utility_PQLAQuadrature.cpp:47]: If there is a difference it's probably immeasurable.  I think either way something gets pushed onto the stack...

**Comment by psbritt** [comment]
I've changed it back to pass by value for native types. @gonuke @bam241, I remember talking about not merging your own PR - would either of you merge this when you get a chance, please?

**Comment by gonuke** [comment]
Thanks for all the work on this @psbritt
